4.5 - Error page refresh

markstory opened the pull request:
Refresh the design of the development error pages. I find the current design pretty busy and hard to navigate, as you can only look at a single frame at a time. The new design has a few improvements:

* Responsive design. The new layout renders reasonably well on tablets and ok on phones. The previous design was a mess on small viewports.
* Focus on the stack trace. The stack trace navigation and context for each frame are now together.
* You can view multiple frames at a time. Because each frame can be collapsed individually you can now look at multiple frames at the same time.
* Chained exceptions are rendered!
* Vendor frames are dimmed. I've removed the 'toggle vendor frames' behavior as I don't think it is that useful.

### Next steps

Next, I'd like to de-duplicate the stack frame rendering so that chained exceptions are easier to read.

### Backwards compatibility

I've changed a good amount of the CSS in an incompatible way. This could cause userland custom error pages to render poorly, but I'm not overly concerned about that. My thinking is that if a user has modified the templates for error pages they hopefully have duplicated and modified the existing error CSS.

I've left the elements used in the previous implementation behind so that if userland code is using them those error pages will continue to work as they always have.

### Screenshots

Please ignore the cake icon on the right side as it is from debug kit.

**Missing Controller** an example of a 'helpful' error message from the framework.

![Screenshot 2022-09-18 at 13-27-21 Error Missing Controller](https://user-images.githubusercontent.com/24086/190920631-31962f18-b8b7-4ea2-8028-ef73e8ae7d8b.png)

**Chained exception rendering**

![Screenshot 2022-09-18 at 13-27-39 Error Page is missing](https://user-images.githubusercontent.com/24086/190920632-b39e7996-f579-4aad-8bb0-cbd279d74b1f.png)

markstory committed:
Refresh development error page

- Simplify page layout and only have the trace.
- Make frames collapsible.
- Make it possible to have multiple frames open at the same time.
- Render chained exceptions well.

markstory committed:
Improve styling of stacktrace and improve responsive layout support.

markstory committed:
Improve design more.

markstory committed:
Remove vendor-frame handling.

ndm2 reviewed (commented):
Can we make the app frame text bold maybe? I find them hard to spot despite the dimming.

On a wider screen it's hard to figure the correct button for the frame you'd like to expand. Maybe a hover effect could help? Having the frame be the toggle itself would probably be even easier to navigate, but I guess it would make selecting text for copying rather cumbersome, and mess up the edit button usage. Another alternative could be to put the toggle on the left.

markstory commented:
> Can we make the app frame text bold maybe? I find them hard to spot despite the dimming.

Do you think a different background for vendor frames would help?

> On a wider screen it's hard to figure the correct button for the frame you'd like to expand. Maybe a hover effect could help? Having the frame be the toggle itself would probably be even easier to navigate, but I guess it would make selecting text for copying rather cumbersome, and mess up the edit button usage. Another alternative could be to put the toggle on the left.

Having the button on the left sounds like a simple solution that places all of the controls/information together. I'll give that a try.

ndm2 commented:
> Do you think a different background for vendor frames would help?

That could work. There might not be much room for further dimming though before things turn into white, so maybe rather app frames should have a darker color, but that's the same outcome I guess, the vendor frames would have a different color anyways 🙃

markstory committed:
Improve contrast for vendor frames and regular frames.

Move the collapse arrow to the left so its relation to the line text is
more obvious.

markstory commented:
![Screenshot 2022-09-20 at 22-11-27 Error Page is missing](https://user-images.githubusercontent.com/24086/191399818-8da80a8e-8227-418d-8c88-68c05cb8894f.png)

What do you think of this @ndm2? I've moved the collapsing arrows to the left and reduced the contrast with the vendor frames more.

ndm2 commented:
Feels much better 👍

ndm2 commented on templates/layout/dev_error.php:
```diff
@@ -218,14 +228,55 @@
+    .stack-frame-edit {
+        margin: 0 5px 0 0;
+    }
+    .stack-frame-toggle {
```
```suggestion
    .stack-frame-toggle {
        cursor: pointer;
```

markstory committed:
Update templates/layout/dev_error.php

Co-authored-by: Oliver Nowak <ndm2@users.noreply.github.com>
